### PR TITLE
MSBot and .bot file docs

### PR DESCRIPTION
**Pull request by vishwacsena**
- Re-wrote MS Bot docs. 
- Added several new sections that demonstrate how MSBot and other CLI tools work together. 
- Docs on .bot file, encryption and other relevant topics.

## MSBot functionality
- [Create](./docs/create-bot.md) a .bot file
- [Encrypting keys](./docs/bot-file-encryption.md) in .bot file
- [Add services](./docs/add-services.md) to .bot file
- [List connected services](./docs/list-connected-services.md) in a .bot file
- [Remove connected service](./docs/remove-service.md) from .bot file
- Using MSBot CLI in conjunction with other CLIs
    - [MSBot and LUIS CLI](./docs/msbot-luis.md)
    - [MSBot and QnA Maker CLI](./docs/msbot-qnamaker.md)
    - [MSBot and Dispatch CLI](./docs/msbot-dispatch.md)
    - [MSBot and AZ CLI](./docs/msbot-az.md)
- [Exporting and cloning](./docs/export-clone.md) a bot

See [here](./docs/botframework-configuration.md) to learn about consuming the .bot file in your bots built with the [Bot Builder SDK](microsoft/botbuilder). This will enable a seamless end-to-end creating, management and consumption of connected services your bot relies on.

See [here](https://github.com/microsoft/botframework-emulator) to learn more about Bot Framework Emulator V4. Emulator also relies on .bot file and provides rich UI based experience to connect and manage connected services your bot relies on.

### CODE REVIEW

File: packages/MSBot/docs/add-services.md [hunk 1]
```diff
@@ -0,0 +1,284 @@
+## Connecting Your Bot to a Service
+
```
Comment (Kaiqb) [inline packages/MSBot/docs/add-services.md]: We need to have examples of the usage. It's hard to map the description of the options with what I see in the Azure portal. For example, if I am using CosmosDB, what does service name mean and how can I find it in the Azure portal?

File: packages/MSBot/docs/add-services.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+```shell
+msbot connect [command]
+```
+
+Where the command is one of the services
+
+| Command                                                 | Description                              |
+|---------------------------------------------------------|------------------------------------------|
+| [appinsights](#connecting-to-azure-appinsights-service) | connect to Azure AppInsights             |
```
Comment (Kaiqb) [inline packages/MSBot/docs/add-services.md]: This list list is the same as I get when I use the help command. I'm not sure if this is adding any value.

File: packages/MSBot/docs/bot-file-encryption.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# Bot Secrets
+
+It is useful for tools like the emulator to have secure access to keys it needs to work with the services that are connected to the bot.  The MSBot tool supports encrypting keys in your .bot file. When you request the .bot file be encrypted, a secret is generated using AES256 and provided to you. 
+
+Any command which accepts the `--secret` option has data which needs to be encrypted with the secret.
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file-encryption.md]: Any command... not clear which command? It needs a bit more context here.

File: packages/MSBot/docs/bot-file.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+Bots usually consume lots of diverse services, such as [LUIS.ai](https://luis.ai), or [QnaMaker.ai](https://qnamaker.ai). When you are developing a bot, there is no uniform place to store the the metadata about the services that are in use.  This prevents us from building tooling that looks at a bot as a whole.
+
+To address this problem, we have created a **.bot file** to act as the place to unify all of these services together to enable tooling.  For example, the [Bot Framework Emulator V4](https://github.com/Microsoft/BotFramework-Emulator/releases) uses a the .bot file to create a unified view over the connected services your bot consumes.  
+
+Via the .bot file, you can register services like:
+
+* **Localhost** local debugger endpoints
+* [**Azure Bot Service**](https://azure.microsoft.com/en-us/services/bot-service/) Azure Bot Service registrations.
+* [**LUIS.AI**](https://www.luis.ai/) Language Understanding (LUIS) allows your application to understand what a person wants in their own words.
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file.md]: LUIS gives your bot the ability to communicate with people using natural language.

File: packages/MSBot/docs/bot-file.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+To address this problem, we have created a **.bot file** to act as the place to unify all of these services together to enable tooling.  For example, the [Bot Framework Emulator V4](https://github.com/Microsoft/BotFramework-Emulator/releases) uses a the .bot file to create a unified view over the connected services your bot consumes.  
+
+Via the .bot file, you can register services like:
+
+* **Localhost** local debugger endpoints
+* [**Azure Bot Service**](https://azure.microsoft.com/en-us/services/bot-service/) Azure Bot Service registrations.
+* [**LUIS.AI**](https://www.luis.ai/) Language Understanding (LUIS) allows your application to understand what a person wants in their own words. 
+* [**QnA Maker**](https://qnamaker.ai/) Build, train and publish a simple question and answer bot based on FAQ URLs, structured documents or editorial content in minutes.
+* [**Dispatch**](https://github.com/Microsoft/botbuilder-tools/tree/master/Dispatch) models for dispatching across heterogeneous sources.
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file.md]: What does heterogeneous mean? A simpler thing to say would be across multiple services.

File: packages/MSBot/docs/bot-file.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+* [**Azure Blob Storage**](https://azure.microsoft.com/en-us/services/storage/blobs/) for bot state persistence. 
+* [**Azure Cosmos DB**](https://azure.microsoft.com/en-us/services/cosmos-db/) - globally distributed, multi-model database service to persist bot state.
+
+Apart from these, your bot might rely on other custom services. You can leverage the [generic service](./add-services.md) capability to connect a generic service configuration.
+
+## When is a .bot file created? 
+- If you create a bot using [Azure Bot Service](https://ms.portal.azure.com/#blade/Microsoft_Azure_Marketplace/GalleryResultsListBlade/selectedSubMenuItemId/%7B%22menuItemId%22%3A%22gallery%2FCognitiveServices_MP%2FBotService%22%2C%22resourceGroupId%22%3A%22%22%2C%22resourceGroupLocation%22%3A%22%22%2C%22dontDiscardJourney%22%3Afalse%2C%22launchingContext%22%3A%7B%22source%22%3A%5B%22GalleryFeaturedMenuItemPart%22%5D%2C%22menuItemId%22%3A%22CognitiveServices_MP%22%2C%22subMenuItemId%22%3A%22BotService%22%7D%7D), a .bot file is automatically created for you with list of connected services provisioned. The .bot is [encrypted](./bot-file-encryption.md) by default.
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file.md]: How do I decrypt the .bot file that was generated when I created the bot in ABS?

File: packages/MSBot/docs/bot-file.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+## What does a bot file look like? 
+[Here's](./sample-bot-file.json) an example .bot file.
+
+## Why .bot file?
+
+- Do I need to use the .bot file to build a bot? 
+- Why do I need the bot file? 
+
+.bot file is **not** a requirement to build amazing bots with [Bot Builder SDK](microsoft/botbuilder).
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file.md]: drop the word amazing

File: packages/MSBot/docs/bot-file.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+Also, you can continue to use appsettings/ web.config/ env, keyvault or any mechanism they you see fit to keep track of service references and keys that your bot depend on. 
+
+The .bot file however addresses few gaps with a bring-your-own approach:
```
Comment (Kaiqb) [inline packages/MSBot/docs/bot-file.md]: The advantages of using .bot file are:

File: packages/MSBot/docs/botframework-configuration.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Using .bot file in your [Bot Builder SDK](microsoft/botbuilder) bot
+Apart from the [Bot Builder CLI tools](../../../README.MD) and the [Bot Framework Emulator V4](microsoft/botframework-emulator) working well with the .bot file, you can also use the .bot file to get service configuration information in your bot's code. 
+
+The BotFramework-Configuration library available for [C#](https://www.nuget.org/packages/Microsoft.Bot.Configuration) and [JS](https://www.npmjs.com/package/botframework-configuration) helps you load a bot file and supports several methods to query and get the appropriate service configuration information. 
+
+You can refer to the [samples](microsoft/botbuilder-samples) to see how the .bot file can be used in your bot's code to retrieve connected service configurations from the .bot file. 
+
+**NOTE:** The bot configuration library would need the encryption key if your bot file is encrypted. You can use the standard mechanisms available for the platform you are on to securely make the key available to your bot. e.g. using application settings for Azure Bot Service bots, using appsettings.json for C#, .env file for Node.js etc.
```
Comment (Kaiqb) [inline packages/MSBot/docs/botframework-configuration.md]: If I am using .appsettings to store the key to the .bot file, then essentially, I am in the same situation that I was in when I used to save all Ids/Pwds in .appsettings, don't I?

File: packages/MSBot/docs/export-clone.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
+## Exporting Services
```
Comment (Kaiqb) [inline packages/MSBot/docs/export-clone.md]: What does check-in resources mean?

File: packages/MSBot/docs/export-clone.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+You can export all of the resources necessary for someone to recreate in their own
+subscription by using the **msbot export services** command.  This is useful for you to
+be able to check-in resources without the keys. 
+
+The result of the command is a folder with all of the QnAMaker, and LUIS models exported
```
Comment (Kaiqb) [inline packages/MSBot/docs/export-clone.md]: The result of the command is a folder with QnAMaker and LUIS models exported...

File: packages/MSBot/docs/msbot-az.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# MSBot and Azure CLI
+Several of the [BotBuilder tools](../../../README.md) are designed to work together. This topic illustrates how MSBot CLI can work in conjunction with the [Azure CLI](../../../AzureCli).
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-az.md]: This topic shows how the MSBot CLI..

File: packages/MSBot/docs/msbot-dispatch.md [hunk 13]
```diff
@@ -0,0 +1,8 @@
+# MSBot and Dispatch CLI
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-dispatch.md]: topic shows

File: packages/MSBot/docs/msbot-dispatch.md [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+Several of the [BotBuilder tools](../../../README.md) are designed to work together. This topic illustrates how MSBot CLI can work in conjunction with the [Dispatch CLI](../../Dispatch).
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-dispatch.md]: topic shows

File: packages/MSBot/docs/msbot-luis.md [hunk 15]
```diff
@@ -0,0 +1,38 @@
+# MSBot and LUIS CLI
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-luis.md]: topic shows

File: packages/MSBot/docs/msbot-luis.md [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+Several of the [BotBuilder tools](../../../README.md) are designed to work together. This topic illustrates how MSBot CLI can work in conjunction with the [LUIS CLI](../../LUIS).
+
+### Create and add a new LUIS application to .bot file
+```bash
+luis add application --in <application_description.json> --msbot | msbot connect luis --stdin [--secret <YOUR-SECRET>]
+```
+
+### Create a new LUIS application by importing a model file and add it to .bot file
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-luis.md]: and adding it to .bot file

File: packages/MSBot/docs/msbot-luis.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+```bash
+luis import application --in <application.json> --msbot | msbot connect luis --stdin [--secret <YOUR-SECRET>]
+```
+
+**Note:** You either do luis add application followed by luis import version .OR. luis import application. luis import application combines add application and import version into one command.
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-luis.md]: There are two ways to ....

File: packages/MSBot/docs/msbot-luis.md [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+### Train and publish LUIS application version using LUIS configuration in .bot file
+```bash
+> msbot get <LUIS-SERVICE-NAME-OR-ID> [--secret <YOUR-SECRET>] | luis train version [--wait] --stdin
+
+> msbot get <LUIS-SERVICE-NAME-OR-ID> [--secret <YOUR-SECRET>] | luis publish version --stdin
+```
+
+### Create a new LUIS application version and update .bot file
+This is a complex chaining to pull the current LUIS configuration from .bot file, create a new LUIS application version 0.2 using the existing LUIS keys from .bot file and updating the .bot file with the new 0.2 version information.
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-luis.md]: You can chain multiple commands to ...

File: packages/MSBot/docs/msbot-qnamaker.md [hunk 19]
```diff
@@ -0,0 +1,23 @@
+# MSBot and QnAMaker CLI
```
Comment (Kaiqb) [inline packages/MSBot/docs/msbot-qnamaker.md]: topic shows

File: packages/MSBot/docs/update-service.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
+## Updating Service
+To update a service with new information you can use the msbot update set of commands
+
```
Comment (Kaiqb) [inline packages/MSBot/docs/update-service.md]: What kind of new information? What does MSBot update set of commands mean?

File: packages/MSBot/docs/update-service.md [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+```shell
+msbot update [serviceType] --[requiredArg] [value]
+```
+
+This allows you to rename a service, update keys, etc.
+
```
Comment (Kaiqb) [inline packages/MSBot/docs/update-service.md]: I can get this list from -help. It's better to add actual sample usage here.

File: packages/MSBot/docs/update-service.md [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+| appinsights | update Azure AppInsights             | --serviceName     |
+| blob        | update Azure Blob storage            | --serviceName     |
+| cosmosdb    | update Azure CosmosDB                | --serviceName     |
+| dispatch    | update Dispatch model                | --applicationId   |
+| endpoint    | update endpoint                      | --endpoint        |
+| generic     | update generic service configuration | --url             |
+| luis        | update LUIS application              | --applicationId   |
+| qna         | update QNA service                   | --kbId            |
+
+These commands take the same args as the msbot connect <servicetype> commands, but only some properties can be updated.
```
Comment (Kaiqb) [inline packages/MSBot/docs/update-service.md]: It would be useful to add the list of properties that can't be updated.

File: packages/MSBot/docs/update-service.md [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+For example, you can change the name of a Azure service, but you can't change the serviceName, tenantId or subscriptionId
+as those values are immutable.
+
+The update commands can be used with switches.  Unlike with connect where you are required to pass all switches in only need
+to pass in the unique id for the service and any changed values.
```
Comment (Kaiqb) [inline packages/MSBot/docs/update-service.md]: This is hard to follow...Not sure what these accomplish.

File: packages/MSBot/docs/botframework-configuration.md (hunk 9 above)
Comment (vishwacsena) [reply packages/MSBot/docs/botframework-configuration.md]: You are not. The reason is because if you keep everything in appsettings then we cannot tool with it.

File: packages/MSBot/docs/update-service.md (hunk 21 above)
Comment (vishwacsena) [reply packages/MSBot/docs/update-service.md]: Yes. but the docs list it so you can look at the scope of functionality offered by the tool even if you don't yet have the tool installed.

File: packages/MSBot/docs/update-service.md (hunk 22 above)
Comment (vishwacsena) [reply packages/MSBot/docs/update-service.md]: Yes. An example list of properties is listed right after this line.

File: packages/MSBot/docs/bot-file-encryption.md [hunk 24]
```diff
@@ -0,0 +1,64 @@
+# Bot Secrets
+
+It is useful for tools like the emulator to have secure access to keys it needs to work with the services that are connected to the bot.  The MSBot tool supports encrypting keys in your .bot file. When you request the .bot file be encrypted, a secret is generated using AES256 and provided to you. 
+
+Any command which accepts the `--secret` option has data which needs to be encrypted with the secret. 
+This allows you to check in a .bot file into a public repo safely and only need the secret to unlock all of the keys your bot uses.
+
+**CAUTION::**
+- There are no retrieval mechanisms in place for retrieving a lost secret key. You should use best practices to secure your secret.  **It is strongly encouraged that you DO NOT check it into your source repo or code directly and instead rely on technologies such as Azure Key Vault to securely store it.**
+- You should make sure that you use the same secret when adding all services.
```
[the remark this answers is not in the document]
Comment (vishwacsena) [reply packages/MSBot/docs/bot-file-encryption.md]: There are no retrieval mechanisms if you lose your key which is what the first line of this paragraph says.

File: packages/MSBot/docs/add-services.md (hunk 2 above)
Comment (vishwacsena) [reply packages/MSBot/docs/add-services.md]: I believe the list is useful for someone looking through the docs even before they've installed the tool.

File: packages/MSBot/docs/add-services.md (hunk 1 above)
Comment (vishwacsena) [reply packages/MSBot/docs/add-services.md]: Filed a bug to track this separately. https://github.com/Microsoft/botbuilder-tools/issues/576
